src/Build/Logging/ParallelLogger/ConsoleOutputAligner.cs

```diff
@@ -0,0 +1,150 @@
+        /// For optimization purposes this method uses single <see cref="StringBuilder"/> instance. This makes this method non thread safe.
+        /// Calling side is expected this usage is non-concurrent. This shall nto be an issue as it is expected that writing into Console shall be serialized anyway.
+        /// </remarks>
+        public string AlignConsoleOutput(string message, bool prefixAlreadyWritten, int prefixWidth)
+        {
+            int i = 0;
+            int j = message.IndexOfAny(MSBuildConstants.CrLf);
+            try
+            {
+                StringBuilder sb = _reusedStringBuilder;
+                Debug.Assert(sb.Length == 0, "Something else is using _reusedStringBuilder. Please make sure this instance method is not called concurrently.");
+                sb.Length = 0;
+                // The string contains new lines, treat each new line as a different string to format and send to the console
+                while (j >= 0)
+                {
+                    AlignAndIndentLineOfMessage(sb, prefixAlreadyWritten, prefixWidth, message, i, j - i);
+                    i = j + (message[j] == '\r' ? 2 : 1);
+                    j = i < message.Length ? message.IndexOfAny(MSBuildConstants.CrLf, i) : -1;
+                }
+
+                // Process rest of message
+                AlignAndIndentLineOfMessage(sb, prefixAlreadyWritten, prefixWidth, message, i, message.Length - i);
+
+                return sb.ToString();
+            }
+            finally
+            {
+                // prepare for reuse
+                _reusedStringBuilder.Length = 0;
+            }
+        }
+
+        /// <summary>
+        /// Append aligned and indented message lines into running <see cref="StringBuilder"/>.
+        /// </summary>
+        private void AlignAndIndentLineOfMessage(StringBuilder sb, bool prefixAlreadyWritten, int prefixWidth, string message, int start, int count)
+        {
+            int bufferWidthMinusNewLine = _bufferWidth - 1;
+
+            bool bufferIsLargerThanPrefix = bufferWidthMinusNewLine > prefixWidth;
+            if (_alignMessages && bufferIsLargerThanPrefix && count > 0)
+            {
+                // If the buffer is larger then the prefix information (timestamp and key) then reformat the messages
+
+                // Beginning index of string to be written
+                int index = 0;
+                // Loop until all the string has been sent to the console
+                while (index < count)
+                {
+                    // Position of virtual console cursor.
+                    // By simulating cursor position adjustment for tab characters '\t' we can compute
+                    //   exact numbers of characters from source string to fit into Console.BufferWidth.
+                    int cursor = 0;
+
+                    // Write prefix if needed
+                    if ((!prefixAlreadyWritten || index > 0 || start > 0) && prefixWidth > 0)
```

Comment:
Sorry if this is stupid, but should this be:
```suggestion
                    if (!(prefixAlreadyWritten || index > 0 || start > 0) && prefixWidth > 0)
```
?

Reply:
Prefix is written when has not been already written or it is not the very 1st line of the message.
I believe this line is correct, it also covered in unite tests.
Input message:
```
prefix:line1\nline2\nline3
```
output:
```
prefix:line1
       line2
       line3
```